sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/HadoopFsRelation.scala

```diff
@@ -71,7 +70,13 @@ case class HadoopFsRelation(
 
   override def sizeInBytes: Long = {
     val compressionFactor = sqlContext.conf.fileCompressionFactor
-    (location.sizeInBytes * compressionFactor).toLong
+    val defaultSize = (location.sizeInBytes * compressionFactor).toLong
+    location match {
+      case cfi: CatalogFileIndex if sparkSession.sessionState.conf.fallBackToHdfsForStatsEnabled =>
```

Comment:
We should only fallback to HDFS if the table stats are not available from table metadata. How can we know the table stats are not available here?

Reply:
I checked the code in `DataSource.resolveRelation`, the `CatalogFileIndex` is created as
```
val index = new CatalogFileIndex(
  sparkSession,
  catalogTable.get,
  catalogTable.get.stats.map(_.sizeInBytes.toLong).getOrElse(defaultTableSize))
```

Reply:
Nice catch! Then, shall we compare the `location.sizeInBytes` with `spark.sql.defaultSizeInBytes` in order to check that?

Reply:
SGTM

Reply:
I think `(location.sizeInBytes * compressionFactor).toLong` is enough.

And `(location.sizeInBytes * compressionFactor).toLong` is faster than `CommandUtils.getSizeInBytesFallBackToHdfs(sparkSession, new Path(cfi.table.location), defaultSize)`.

That why I only recalculate table statistics if we go this code path:
https://github.com/apache/spark/blob/c30b5297bc607ae33cc2fcf624b127942154e559/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L383-L387

https://github.com/apache/spark/pull/24715/files#diff-d99813bd5bbc18277e4090475e4944cfR643-R646

Reply:
@wangyum . I'm not sure what you mean by the following. Do you know the goal of this PR was to prevent `8.0EB` situation?
> I think (location.sizeInBytes * compressionFactor).toLong is enough.

Also, ping @shahidki31 since he is the author of this PR.

Reply:
Yes. I have prepared some tests to illustrate this issue. These tests can be passed before this commit:
```scala
  test("Non-partitioned data source table") {
    withTempDir { dir =>
      withTable("spark_25474") {
        sql(s"CREATE TABLE spark_25474 (c1 BIGINT) USING PARQUET LOCATION '${dir.toURI}'")
        spark.range(5).write.mode(SaveMode.Overwrite).parquet(dir.getCanonicalPath)

        assert(getCatalogTable("spark_25474").stats.isEmpty)
        val relation = spark.table("spark_25474").queryExecution.analyzed.children.head
        assert(relation.stats.sizeInBytes === 935)
      }
    }
  }

  test("Partitioned data source table default") {
    withTempDir { dir =>
      withTable("spark_25474") {
        spark.sql("CREATE TABLE spark_25474(a int, b int) USING parquet " +
          s"PARTITIONED BY(a) LOCATION '${dir.toURI}'")
        spark.sql("INSERT INTO TABLE spark_25474 PARTITION(a=1) SELECT 2")

        assert(getCatalogTable("spark_25474").stats.isEmpty)
        val relation = spark.table("spark_25474").queryExecution.analyzed.children.head
        // scalastyle:off line.size.limit
        // It's 8.0EB in this case. This 8.0EB from:
        // https://github.com/apache/spark/blob/c30b5297bc607ae33cc2fcf624b127942154e559/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L383-L387
        // scalastyle:on line.size.limit
        assert(relation.stats.sizeInBytes === conf.defaultSizeInBytes)
      }
    }
  }

  test("Partitioned data source table and disable HIVE_MANAGE_FILESOURCE_PARTITIONS") {
    withSQLConf(SQLConf.HIVE_MANAGE_FILESOURCE_PARTITIONS.key -> "false") {
      withTempDir { dir =>
        withTable("spark_25474") {
          spark.sql("CREATE TABLE spark_25474(a int, b int) USING parquet " +
            s"PARTITIONED BY(a) LOCATION '${dir.toURI}'")
          spark.sql("INSERT INTO TABLE spark_25474 PARTITION(a=1) SELECT 2")

          assert(getCatalogTable("spark_25474").stats.isEmpty)
          val relation = spark.table("spark_25474").queryExecution.analyzed.children.head
          assert(relation.stats.sizeInBytes === 418)
        }
      }
    }
  }
```

https://github.com/apache/spark/compare/master...wangyum:SPARK-25474-DEV?expand=1

Reply:
@wangyum can you send a PR for your proposal? It's unclear to me what you are proposing here.

Reply:
https://github.com/apache/spark/pull/24715

Reply:
I am not sure, is there any issue in this PR. As per this code, if the table doesn't have any statistics, then only will come to the `sizeInBytes` method. May be we can add the extra check mentioned above.
https://github.com/apache/spark/blob/0526529b31737e5bf4829f8259f3a020f2cc51f1/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala#L42-L46

Reply:
IIUC the issue in this PR is, we always fallback to HDFS stats even if table stats are available.

Reply:
@cloud-fan I have created a followup PR to add the extra condition. Thanks

Reply:
`Yes. I have prepared some tests to illustrate this issue. These tests can be passed before this commit:`

@wangyum First and 3rd will pass after the PR. 2nd test is a bug which fixed in the commit. 

Btw, 1st and 3rd tests are not CatalogFileIndex, which any way won't come to this flow